### ISSUE TITLE
Simulcast Egest

### ISSUE THREAD

**Issue by robashton** [webrtc-rs/webrtc#289]
Bit o' a doozy this one

I'd quite like to *send* some Simulcast from webrtc-rs, and at the moment as far as I can tell, I can't even hack my way around this (please do correct me if I'm wrong), the setup of the transceivers/tracks/etc seem to be very much working against me.

I see there was this delightful commit over in Pion a few months ago

https://github.com/pion/webrtc/commit/37e16a3b15a3a474573b88eb7eeb728afaeb3c58

The gist being that we end up with an AddEncoding method on a track, and the rest is mostly worked out for us automatically.

Is there going to be a huge problem with me porting this over to the current version of webrtc-rs, or does this get in the way of the roadmap? I expect it's a bit of a chunk of work, so I'd like to garner some feelings on the matter before I dive in and start slinging code all over the place.

**Comment by k0nserv** [comment]
Happy to have it included, but I'd like it to be implemented according to the specification, just noting in case Pion isn't. I believe that Pion/Webrtc-rs does not implement ingress simulcast correctly currently.

**Comment by robashton** [comment]
It is my my understanding that ingest is also a bit sketchy, yes.

```
 but I'd like it to be implemented according to the specification,
```

I don't think as far as I can see, anybody is presently even uttering the all important

```
a=simulcast
```

At either end of the negotiation at the moment.

I think one of the obstacles to "doing things to spec", is that there is overlap between "doing simulcast/SDP to spec", and "aligning the client APIs to spec", as of course the web browsers currently do simulcast via a largely automatic process (GetParameters, setting encdoers, SetParameters, then let the browser do all of the work). My little two hour survey of the scene just now at least points towards browser support for sending simulcast being a bit of a mess at the moment.

Trying to reconcile that world with these server side APIs is a bit awkward, which probably explains why Pion have gone down the route they have (again, at least as far as I can tell from my spelunking just now)

**Comment by robashton** [comment]
I suspect that what Pion is doing fulfils the "to spec" as far as protocol goes - I haven't dug into the full implementation yet, only the commit linked above so I don't know how or what it ends up generating for the negotiation. I imagine it utters all the right incanctations per spec, because these days the spec for *that* seems pretty well defined.

As far as API goes, I'd largely trust their judgement, our needs are basically that we're able to get an RTP sender for a particular encode, and send RTP over it (as opposed to now, where you have an RTP sender per track/transceiver), this is something that the browsers don't have because its all internal!

**Comment by k0nserv** [comment]
Yes I agree that the API question is a bit confusing since the browser handles it behind the scenes. I haven't looked at this in a while, but what I recall being the problem with ingress simulcast is that webrtc-rs treats a simulcast track with three different encodings as three separate tracks. Here I'd like it to be aligned with the spec i.e. it's still a single track(`on_track` only fires a single time) even though there are three distinct SSRCs(+RTX). Likewise for egress simulcast I think we should match the spec i.e., there's a single egress track, and then have a way to specify when sending which `rid` the RTP packet in question applies for.

**Comment by robashton** [comment]
I guess this work comes in two passes then, which is to make the current code a bit more like the Pion code with respect to the above (a single on_track firing), and then adding the convenience methods to set up multiple encodes on the outbound and filter inbounds by RID when reading.

**Comment by k0nserv** [comment]
It does look like Pion has implemented egress simulcast by having a track per send encoding, ideally we'd stick to the spec and have each `RTCRTPSender` have only a single track, but maybe there are complications with this that I'm not aware of.

> I guess this work comes in two passes then, which is to make the current code a bit more like the Pion code with respect to the above (a single on_track firing)[...]

AFAIK Pion also fires `on_track` N times for N send encodings on the remote side. My beef is with both Pion and Webrtc-rs not following the spec in this instance. If you aren't concerned with ingress at the moment I would keep it out of scope

**Comment by robashton** [comment]
Ah, you've already dived deeper than me then :)

I can live with not following the spec if it is too complicated to make work on first pass, if we agree that morally we *should* be following the spec and work towards that as a goal.

**Comment by robashton** [comment]
I'll not get started on this right away, as I need to work out what the timeline is for us requiring this (likely to be short/medium term), but there may be some other priorities that take immediate precedence.

When I do, my first action will simply be to attempt the Pion changes *as they are* in order to understand what the current gap is between where the webrtc-rs codebase is and where Pion has ended up, and I imagine the result of that exercise will dictate how it is done in practise! 

Will update this thread once I start poking it - cheers.

**Comment by robashton** [comment]
(Just FYI, starting this today, in case anybody else was thinking of looking)

**Comment by haaspors** [comment]
I've submitted https://github.com/webrtc-rs/webrtc/pull/578 which ports several pion patches including the one reference earlier in this issue. My PR was merged earlier this month. So I guess we can close this issue and possibly #312
